server/dive_server/serializers/viame.py

```diff
@@ -159,13 +159,70 @@ def _parse_row_for_tracks(row: List[str]) -> Tuple[Feature, Dict, Dict, List]:
     return feature, attributes, track_attributes, confidence_pairs
 
 
-def load_csv_as_tracks(rows: List[str]) -> Dict[str, dict]:
+def create_attributes(
+    metadata_attributes: Dict[str, Attribute],
+    test_vals: Dict[str, int],
+    atr_type: str,
+    key: str,
+    val,
+):
+    valstring = f'{val}'
+    if f'{atr_type}_{key}' not in metadata_attributes:
+        metadata_attributes[f'{atr_type}_{key}'] = {
+            'belongs': atr_type,
+            'datatype': 'text',
+            'name': key,
+            'key': f'{atr_type}_{key}',
+        }
+        test_vals[f'{atr_type}_{key}'] = {}
+        test_vals[f'{atr_type}_{key}'][valstring] = 1
+    elif (
+        f'{atr_type}_{key}' in metadata_attributes and f'{atr_type}_{key}' in test_vals
+    ):
+        if valstring in test_vals[f'{atr_type}_{key}']:
+            test_vals[f'{atr_type}_{key}'][valstring] += 1
+        else:
+            test_vals[f'{atr_type}_{key}'][valstring] = 1
+
+
+def calculate_attribute_types(
+    metadata_attributes: Dict[str, Attribute], test_vals: Dict[str, int]
+):
+    predefined_min_count = (
+        3  # count all keys must have a value to convert to predefined
+    )
+    for attributeKey in metadata_attributes.keys():
+        if attributeKey in test_vals:
+            attribute_type = 'number'
+            low_count = predefined_min_count
+            values = []
+            for (key, val) in test_vals[attributeKey].items():
+                if val <= low_count:
+                    low_count = val
+                values.append(key)
+                if attribute_type == 'number':
+                    try:
+                        float(key)
+                    except ValueError:
+                        attribute_type = 'boolean'
+                if attribute_type == 'boolean' and key != 'True' and key != 'False':
```

Comment:
According to https://viame.readthedocs.io/en/latest/section_links/detection_file_conversions.html, `true` and `false` are valid boolean values.

Reply:
These are a string representation of the values used within the tracks/detections.  I build up a dictionary with the values as a  key to determine how many times a specific value is used.  It's part of trying to see if there are predefined values for text strings (Not the best way to do this but hopefully can talk about it in the meeting today).  Here I start with number type, then see if any of the values-converted-to-string are not boolean to kick it over to text type.  So using `true` and `false` in the CSV will properly convert to `'True'` and `'False'` here.  I think the change would be if we wanted to support `True` and `False` directly in the CSV instead of `true` and `false`.

Reply:
So parsing converts the strings to values (bool, float, str), create_attributes converts them back to strings, and calculate_attribute_types converts them back to primitives again to test their type.

Out of curiosity, why convert to string and back at all?  int, float, and bool are all valid dictionary keys?

Then you could just `type(key) is bool`

Reply:
Another possible alternative is to just add all the values to an array from create_attributes, then generate your Dict[value, count] using `collections.Counter` https://docs.python.org/3/library/collections.html#collections.Counter

Reply:
I'm guessing tunnel vision from doing this in electron first and copying it over.

Reply:
Hmm.  Makes sense.  There's arguably value in having identical logic in python and js (I made that argument at some point I think).  If you want to keep this as-is, I think test coverage mostly takes care of my concerns.